rtos/ConditionVariable.cpp

```diff
@@ -0,0 +1,142 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "ConditionVariable.h"
+#include "Semaphore.h"
+
+#include "mbed_error.h"
+#include "mbed_assert.h"
+
+namespace rtos {
+
+#define RESUME_SIGNAL      (1 << 15)
+
+struct Waiter {
+    Waiter();
+    Semaphore sem;
+    Waiter *prev;
+    Waiter *next;
+    bool in_list;
+};
+
+Waiter::Waiter(): sem(0), prev(NULL), next(NULL), in_list(false)
+{
+    // No initialization to do
+}
+
+ConditionVariable::ConditionVariable(Mutex &mutex): _mutex(mutex), _wait_list(NULL)
+{
+    // No initialization to do
+}
+
+void ConditionVariable::wait()
+{
+    wait_for(osWaitForever);
+}
+
+bool ConditionVariable::wait_for(uint32_t millisec)
+{
+    Waiter current_thread;
+    MBED_ASSERT(_mutex.is_owner());
+    MBED_ASSERT(_mutex.count == 1);
+    _add_wait_list(&current_thread);
+
+    // Unlock mutex
+    uint32_t count = _mutex.count;
```

Comment:
I left the code to fallback which unlocks the mutex if this condition occurs in a release build (asserts disabled). The alternatives would be to trap in release builds as well or to leave the mutex held and deadlock (since this condition variable can never be singled again and the mutex could never be released).

Reply:
I think either trap or deadlock would be more appropriate, and I'd be happy with deadlock. (Especially if it meant not needing the added re-count overhead on Mutex).

It's a programming error, not a run-time error. It can fail equally on all builds.

If it did deadlocks on a release build, switching to debug would immediately tell you why due to the assert.

I don't see any merit in adding special code only relevant to the release build to make it limp on, despite the programming error. The logic is normally that the release build doesn't have the programming errors because they've all been asserted out by the debug builds.

Reply:
Makes sense. I removed the multiple unlocks